### PR TITLE
Use Micrometer metrics instead of in memory one

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
Since I'm no metrics expert I need feedback from such as experts as @shakuzen @adriancole @msmsimondean and the rest of the experts interested in this topic ;)

I've added an initial implementation that uses `counter`s and `gauge`s from micrometer for metrics. I've passed some default values for metric names but I'm not sure if those are the best.

This would actually start defaulting to the Micrometer version of a metric reporter since it's pretty much always on the classpath. If someone doesn't have a `meterRegistry` bean, then the in memory option will be taken into account.

I don't think we can have an option that there is no micrometer on the classpath since it comes with boot.

UPDATE: As @spencergibb mentioned, micrometer comes as an dependency with an actuator, however actuator doesn't always have to be there. As @adriancole mentioned there already is a zipkin micrometer reporter that we can reuse.

### CODE REVIEW

**Comment by msmsimondean** [comment]
@marcingrzejszczak would it be worth using the Noop metrics reporter if `meterRegistry` bean is not registered.  What do you think @adriancole?

**Comment by spencergibb** [comment]
Micrometer is optional in boot, it is pulled in with the actuator starter.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java [hunk 1]
```diff
@@ -283,4 +274,106 @@ public String toString() {
...
+		@ConditionalOnMissingBean
+		ReporterMetrics sleuthReporterMetrics() {
+			return new InMemoryReporterMetrics();
+		}
+
+	}
+
+	@Configuration
+	@ConditionalOnBean(MeterRegistry.class)
+	static class TraceMicrometerMetricsConfiguration {
```
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java]: This should be a toplevel auto configuration conditional on a micrometer class

**Comment by marcingrzejszczak** [comment]
> @marcingrzejszczak would it be worth using the Noop metrics reporter if meterRegistry bean is not registered. What do you think @adriancole?

that's already the case

> Micrometer is optional in boot, it is pulled in with the actuator starter.

@spencergibb I wasn't sure of that, initially I did exactly what you suggested (with class check) and then I removed it. I'll bring it back.

**Comment by adriancole** [comment]
is there a reason we aren't using this? if so, maybe we mention https://github.com/openzipkin/zipkin-reporter-java/tree/master/metrics-micrometer

**Comment by marcingrzejszczak** [comment]
Most likely cause I didn't know that it's a thing

File: spring-cloud-sleuth-core/pom.xml [hunk 2]
```diff
@@ -286,20 +289,28 @@
 			<artifactId>brave-instrumentation-grpc</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>io.zipkin.reporter2</groupId>
+			<artifactId>zipkin-reporter-metrics-micrometer</artifactId>
+			<exclusions>
+				<exclusion>
+					<groupId>io.micrometer</groupId>
```
Comment (adriancole) [inline spring-cloud-sleuth-core/pom.xml]: @shakuzen reminds me.. we probably should have had this set to "provided" scope in upstream anyway

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java [hunk 3]
```diff
@@ -283,4 +282,40 @@ public String toString() {
 
 	}
 
+	@Configuration
+	@ConditionalOnMissingClass("io.micrometer.core.instrument.MeterRegistry")
+	static class TraceMetricsNoOpConfiguration {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java]: should this be

```suggestion
	static class TraceMetricsInMemoryConfiguration {
```

if no-op is desired, that would be different and cheaper.

**Pull request merged**